File: exercise_HashTable.py
#A kidnapper wrote a ransom note but is worried it will be traced back to him. He found a magazine and wants to know if he can cut out whole words from it and use them to create an untraceable replica of his ransom note. The words in his note are case-sensitive and he must use whole words available in the magazine, meaning he cannot use substrings or concatenation to create the words he needs.

#Given the words in the magazine and the words in the ransom note, print Yes if he can replicate his ransom note exactly using whole words from the magazine; otherwise, print No.

#Input Format

#The first line contains two space-separated integers describing the respective values of  (the number of words in the magazine) and  (the number of words in the ransom note).
#The second line contains  space-separated strings denoting the words present in the magazine.
#The third line contains  space-separated strings denoting the words present in the ransom note.

#Constraints
#Each word consists of English alphabetic letters. The words in the note and magazine are case-sensitive.

#Output format
#Print Yes if he can use the magazine to create an untraceable replica of his ransom note; otherwise, print No.

#Sample Input
#6 4
#give me one grand today night
#give one grand today

#Sample Output
#Yes

def ransom_note(magazine, ransom):
  wordinMagazine = dict()
    wordinRansome = dict()
    
    for word in magazine:
      if word not in wordinMagazine:
        wordinMagazine[word] = 0
        wordinMagazine[word] = wordinMagazine[word] + 1
    for word in ransom:
      if word not in wordinMagazine:
        return False
        if wordinMagazine[word] < 1:
          return False
        wordinMagazine[word] = wordinMagazine[word] - 1
    return True

m, n = map(int, input().strip().split(' '))
magazine = input().strip().split(' ')
ransom = input().strip().split(' ')
answer = ransom_note(magazine, ransom)
if(answer):
  print("Yes")
else:
  print("No")

File: exercise_HashTable.cpp
//Used C++ STL container unordered_map to solve a Hackerrank/CTCI hashtable problem

/*
A kidnapper wrote a ransom note but is worried it will be traced back to him. He found a magazine and wants to know if he can cut out whole words from it and use them to create an untraceable replica of his ransom note. The words in his note are case-sensitive and he must use whole words available in the magazine, meaning he cannot use substrings or concatenation to create the words he needs.

Given the words in the magazine and the words in the ransom note, print Yes if he can replicate his ransom note exactly using whole words from the magazine; otherwise, print No.

Input Format

The first line contains two space-separated integers describing the respective values of  (the number of words in the magazine) and  (the number of words in the ransom note).
The second line contains  space-separated strings denoting the words present in the magazine.
The third line contains  space-separated strings denoting the words present in the ransom note.

Constraints


.
Each word consists of English alphabetic letters (i.e.,  to  and  to ).
The words in the note and magazine are case-sensitive.
Output Format

Print Yes if he can use the magazine to create an untraceable replica of his ransom note; otherwise, print No.

Sample Input

6 4
give me one grand today night
give one grand today
Sample Output

Yes
Explanation

All four words needed to write an untraceable replica of the ransom note are present in the magazine, so we print Yes as our answer.
 */

#include <map>
#include <set>
#include <list>
#include <cmath>
#include <ctime>
#include <deque>
#include <queue>
#include <stack>
#include <string>
#include <bitset>
#include <cstdio>
#include <limits>
#include <vector>
#include <climits>
#include <cstring>
#include <cstdlib>
#include <fstream>
#include <numeric>
#include <sstream>
#include <iostream>
#include <algorithm>
#include <unordered_map>

using namespace std;

bool ransom_note(vector<string> magazine, vector<string> ransom) {
    unordered_map <string, int> wordsAvailable;
    
    int num_wordsinMagazine = magazine.size();
    int num_wordsinRansom = ransom.size();
    
    for(int i = 0; i < num_wordsinMagazine; i++)
        wordsAvailable[magazine[i]]++;
    
    for(int i = 0; i < num_wordsinRansom; i++){
        if(!wordsAvailable[ransom[i]])
            return false;
        wordsAvailable[ransom[i]]--;
    }
    
    return true;
}

int main(){
    int m;
    int n;
    cin >> m >> n;
    vector<string> magazine(m);
    for(int magazine_i = 0;magazine_i < m;magazine_i++){
       cin >> magazine[magazine_i];
    }
    vector<string> ransom(n);
    for(int ransom_i = 0;ransom_i < n;ransom_i++){
       cin >> ransom[ransom_i];
    }
    if(ransom_note(magazine, ransom))
        cout << "Yes\n";
    else
        cout << "No\n";
    return 0;
}
